### ISSUE TITLE
Behavioral changes in 1.4.x?

### ISSUE THREAD

**Issue by mayou36** [scikit-hep/iminuit#446]
I've observed a different behavior of the minimizer in 1.3.x to 1.4.x, but did not find anything in the Changelog, but maybe I missed it. Thanks a lot for the updates!

I have a fit where I provide the gradient. For some reason, it does not converge well with it (happened before, fine, maybe the gradient is wrong...). But in 1.3.x minuit failed (spectacularly) at some point, while in 1.4.x, it becomes stuck in an endless loop.

Here is the output. Basically, it seems to try to find a minimum a few times (3 here) and then gets stuck somehow (which is repeated indefinitely). As a sidenote, if an NaN occurred in the loss, the previous loss with a penalty is returned.

Did some behavior change interally? Or was this a bug before?

OLD BEHAVIOR (1.3.x):
```
VariableMetric: start iterating until Edm is < 0.001
VariableMetric: Initial state   - FCN =   4470682.788238 Edm =  1.47635e+08 NCalls =      1
VariableMetric: Iteration #   0 - FCN =   4470682.788238 Edm =  1.47635e+08 NCalls =      1
VariableMetric: Iteration #   1 - FCN =    4451508.55971 Edm =  2.16974e+07 NCalls =     12
VariableMetric: Iteration #   2 - FCN =   4421005.175532 Edm =  4.82029e+06 NCalls =     22
VariableMetric: Iteration #   3 - FCN =   4371448.837892 Edm =       164128 NCalls =     26
VariableMetric: Iteration #   4 - FCN =   4349731.808529 Edm =       148345 NCalls =     31
VariableMetric: Iteration #   5 - FCN =   4079484.078334 Edm =      26763.9 NCalls =     34
VariableMetric: Iteration #   6 - FCN =   3707688.098757 Edm =       295316 NCalls =     39
VariableMetric: Iteration #   7 - FCN =   3621795.764204 Edm =       409895 NCalls =     43
VariableMetric: Iteration #   8 - FCN =   3621795.764204 Edm =       409895 NCalls =     54
VariableMetric: After Hessian   - FCN =   3621795.764204 Edm =       409895 NCalls =    133
VariableMetric: Iteration #   9 - FCN =   3621795.764204 Edm =       409895 NCalls =    133
VariableMetric: Iteration #  10 - FCN =   3621795.764204 Edm =       409895 NCalls =    144
------------------------------------------------------------------
| FCN = 3.622E+06               |     Ncalls=133 (144 total)     |
| EDM = 4.1E+05 (Goal: 0.001)   |            up = 0.5            |
------------------------------------------------------------------
|  Valid Min.   | Valid Param.  | Above EDM | Reached call limit |
------------------------------------------------------------------
|     False     |     True      |   True    |       False        |
------------------------------------------------------------------
| Hesse failed  |   Has cov.    | Accurate  | Pos. def. | Forced |
------------------------------------------------------------------
|     True      |     True      |   False   |   False   | False  |
------------------------------------------------------------------
```

NEW BEHAVIOR (1.4.x)

```
VariableMetric: start iterating until Edm is < 0.001
VariableMetric: Initial state   - FCN =   4470682.788238 Edm =  1.47635e+08 NCalls =      1
VariableMetric: Iteration #   0 - FCN =   4470682.788238 Edm =  1.47635e+08 NCalls =      1
VariableMetric: Iteration #   1 - FCN =    4451508.55971 Edm =  2.16974e+07 NCalls =     12
VariableMetric: Iteration #   2 - FCN =   4421005.175532 Edm =  4.82029e+06 NCalls =     22
VariableMetric: Iteration #   3 - FCN =   4371448.837892 Edm =       164128 NCalls =     26
VariableMetric: Iteration #   4 - FCN =   4349731.808529 Edm =       148345 NCalls =     31
VariableMetric: Iteration #   5 - FCN =   4079484.078334 Edm =      26763.9 NCalls =     34
VariableMetric: Iteration #   6 - FCN =   3707688.098757 Edm =       295316 NCalls =     39
VariableMetric: Iteration #   7 - FCN =   3621795.764204 Edm =       409895 NCalls =     43
Info: DavidonErrorUpdator: delgam < 0 : first derivatives increasing along search line
Info: VariableMetricBuilder: matrix not pos.def, gdel > 0
Info: gdel = 463410
Info in negative or zero diagonal element in covariance matrix : i = 1
Info in negative or zero diagonal element in covariance matrix : i = 4
Info in negative or zero diagonal element in covariance matrix : i = 5
Info in negative or zero diagonal element in covariance matrix : i = 12
Info in added to diagonal of Error matrix a value : dg = 0.500002
Info: gdel = -4.23523e+13
Info: VariableMetricBuilder: no improvement in line search
Info: VariableMetricBuilder: iterations finish without convergence.
Info in VariableMetricBuilder : edm = 1.09721e+06
Info in             requested : edmval = 0.001
VariableMetric: Iteration #   8 - FCN =   3621795.764204 Edm =       409895 NCalls =     54
VariableMetric: After Hessian   - FCN =   3621795.764204 Edm =       409895 NCalls =    133
VariableMetric: Iteration #   9 - FCN =   3621795.764204 Edm =       409895 NCalls =    133
Info in MnHesse: 2nd derivative zero for Parameter  : name = Phi_a_u
Info: MnHesse fails and will return diagonal matrix 
Info: VariableMetricBuilder: Tolerance is not sufficient, continue the minimization
Info in Current  Edm is : edm = 409895
Info in Required Edm is : edmval = 0.001
Info: VariableMetricBuilder: no improvement in line search
Info: VariableMetricBuilder: iterations finish without convergence.
Info in VariableMetricBuilder : edm = 1.63958e+06
Info in             requested : edmval = 0.001
Info: VariableMetricBuilder: FunctionMinimum is invalid after second try
VariableMetric: Iteration #  10 - FCN =   3621795.764204 Edm =       409895 NCalls =    144
VariableMetric: start iterating until Edm is < 0.001
VariableMetric: Initial state   - FCN =   3621795.764204 Edm =      0.50998 NCalls =      1
VariableMetric: Iteration #   0 - FCN =   3621795.764204 Edm =      0.50998 NCalls =      1
VariableMetric: Iteration #   1 - FCN =   3620773.211262 Edm =     0.508285 NCalls =      8
VariableMetric: Iteration #   2 - FCN =   3387012.874856 Edm =      93270.2 NCalls =     10
VariableMetric: Iteration #   3 - FCN =   3340307.716837 Edm =      66423.7 NCalls =     13
VariableMetric: Iteration #   4 - FCN =   3321298.996471 Edm =      64068.5 NCalls =     17
VariableMetric: Iteration #   5 - FCN =   3305552.066694 Edm =      53162.3 NCalls =     21
VariableMetric: Iteration #   6 - FCN =   3305450.398824 Edm =        56735 NCalls =     32
VariableMetric: Iteration #   7 - FCN =   3305446.186449 Edm =      58682.5 NCalls =     42
VariableMetric: Iteration #   8 - FCN =    3305441.97251 Edm =      58690.3 NCalls =     52
VariableMetric: Iteration #   9 - FCN =   3305437.758588 Edm =      58690.2 NCalls =     62
VariableMetric: Iteration #  10 - FCN =   3305433.544525 Edm =      58690.9 NCalls =     72
VariableMetric: Iteration #  11 - FCN =    3305429.33032 Edm =      58691.6 NCalls =     82
VariableMetric: Iteration #  12 - FCN =   3305425.116833 Edm =        58688 NCalls =     92
VariableMetric: Iteration #  13 - FCN =   3305420.903273 Edm =      58688.4 NCalls =    102
VariableMetric: Iteration #  14 - FCN =   3305420.692608 Edm =      58690.7 NCalls =    113
VariableMetric: Iteration #  15 - FCN =   3305420.481937 Edm =      58691.3 NCalls =    124
VariableMetric: Iteration #  16 - FCN =   3305420.271256 Edm =      58692.3 NCalls =    135
VariableMetric: Iteration #  17 - FCN =    3305420.06058 Edm =      58691.9 NCalls =    146
VariableMetric: Iteration #  18 - FCN =    3305420.06058 Edm =      58691.9 NCalls =    157
Info: VariableMetricBuilder: no improvement in line search
Info: VariableMetricBuilder: iterations finish without convergence.
Info in VariableMetricBuilder : edm = 58704.4
Info in             requested : edmval = 0.001
Info in VariableMetricBuilder: INVALID function minimum - edm is above tolerance, : edm = 58691.9
Info in VariableMetricBuilder: Required tolerance  is 10 x edmval  : edmval = 0.001
VariableMetric: start iterating until Edm is < 0.001
VariableMetric: Initial state   - FCN =    3305420.06058 Edm =      58689.9 NCalls =      1
VariableMetric: Iteration #   0 - FCN =    3305420.06058 Edm =      58689.9 NCalls =      1
Info: VariableMetricBuilder: no improvement in line search
Info: VariableMetricBuilder: iterations finish without convergence.
Info in VariableMetricBuilder : edm = 58689.9
Info in             requested : edmval = 0.001
Info in VariableMetricBuilder: INVALID function minimum - edm is above tolerance, : edm = 58689.9
Info in VariableMetricBuilder: Required tolerance  is 10 x edmval  : edmval = 0.001
VariableMetric: Iteration #   1 - FCN =    3305420.06058 Edm =      58689.9 NCalls =     12
VariableMetric: start iterating until Edm is < 0.001
VariableMetric: Initial state   - FCN =    3305420.06058 Edm =      58689.9 NCalls =      1
VariableMetric: Iteration #   0 - FCN =    3305420.06058 Edm =      58689.9 NCalls =      1
VariableMetric: Iteration #   1 - FCN =    3305420.06058 Edm =      58689.9 NCalls =     12
Info: VariableMetricBuilder: no improvement in line search
Info: VariableMetricBuilder: iterations finish without convergence.
Info in VariableMetricBuilder : edm = 58689.9
Info in             requested : edmval = 0.001
Info in VariableMetricBuilder: INVALID function minimum - edm is above tolerance, : edm = 58689.9
Info in VariableMetricBuilder: Required tolerance  is 10 x edmval  : edmval = 0.001
VariableMetric: start iterating until Edm is < 0.001
VariableMetric: Initial state   - FCN =    3305420.06058 Edm =      58689.9 NCalls =      1
VariableMetric: Iteration #   0 - FCN =    3305420.06058 Edm =      58689.9 NCalls =      1
VariableMetric: Iteration #   1 - FCN =    3305420.06058 Edm =      58689.9 NCalls =     12
Info: VariableMetricBuilder: no improvement in line search
Info: VariableMetricBuilder: iterations finish without convergence.
Info in VariableMetricBuilder : edm = 58689.9
Info in             requested : edmval = 0.001
Info in VariableMetricBuilder: INVALID function minimum - edm is above tolerance, : edm = 58689.9
Info in VariableMetricBuilder: Required tolerance  is 10 x edmval  : edmval = 0.001
VariableMetric: start iterating until Edm is < 0.001
VariableMetric: Initial state   - FCN =    3305420.06058 Edm =      58689.9 NCalls =      1
VariableMetric: Iteration #   0 - FCN =    3305420.06058 Edm =      58689.9 NCalls =      1
Info: VariableMetricBuilder: no improvement in line search
Info: VariableMetricBuilder: iterations finish without convergence.
Info in VariableMetricBuilder : edm = 58689.9
Info in             requested : edmval = 0.001
Info in VariableMetricBuilder: INVALID function minimum - edm is above tolerance, : edm = 58689.9
Info in VariableMetricBuilder: Required tolerance  is 10 x edmval  : edmval = 0.001
VariableMetric: Iteration #   1 - FCN =    3305420.06058 Edm =      58689.9 NCalls =     12
VariableMetric: start iterating until Edm is < 0.001
VariableMetric: Initial state   - FCN =    3305420.06058 Edm =      58689.9 NCalls =      1
VariableMetric: Iteration #   0 - FCN =    3305420.06058 Edm =      58689.9 NCalls =      1
Info: VariableMetricBuilder: no improvement in line search
Info: VariableMetricBuilder: iterations finish without convergence.
Info in VariableMetricBuilder : edm = 58689.9
Info in             requested : edmval = 0.001
Info in VariableMetricBuilder: INVALID function minimum - edm is above tolerance, : edm = 58689.9
Info in VariableMetricBuilder: Required tolerance  is 10 x edmval  : edmval = 0.001
VariableMetric: Iteration #   1 - FCN =    3305420.06058 Edm =      58689.9 NCalls =     12
VariableMetric: start iterating until Edm is < 0.001
VariableMetric: Initial state   - FCN =    3305420.06058 Edm =      58689.9 NCalls =      1
VariableMetric: Iteration #   0 - FCN =    3305420.06058 Edm =      58689.9 NCalls =      1
VariableMetric: Iteration #   1 - FCN =    3305420.06058 Edm =      58689.9 NCalls =     12
Info: VariableMetricBuilder: no improvement in line search
Info: VariableMetricBuilder: iterations finish without convergence.
Info in VariableMetricBuilder : edm = 58689.9
Info in             requested : edmval = 0.001
Info in VariableMetricBuilder: INVALID function minimum - edm is above tolerance, : edm = 58689.9
Info in VariableMetricBuilder: Required tolerance  is 10 x edmval  : edmval = 0.001
VariableMetric: start iterating until Edm is < 0.001
VariableMetric: Initial state   - FCN =    3305420.06058 Edm =      58689.9 NCalls =      1
VariableMetric: Iteration #   0 - FCN =    3305420.06058 Edm =      58689.9 NCalls =      1
Info: VariableMetricBuilder: no improvement in line search
Info: VariableMetricBuilder: iterations finish without convergence.
Info in VariableMetricBuilder : edm = 58689.9
Info in             requested : edmval = 0.001
Info in VariableMetricBuilder: INVALID function minimum - edm is above tolerance, : edm = 58689.9
Info in VariableMetricBuilder: Required tolerance  is 10 x edmval  : edmval = 0.001
VariableMetric: Iteration #   1 - FCN =    3305420.06058 Edm =      58689.9 NCalls =     12
VariableMetric: start iterating until Edm is < 0.001
VariableMetric: Initial state   - FCN =    3305420.06058 Edm =      58689.9 NCalls =      1
VariableMetric: Iteration #   0 - FCN =    3305420.06058 Edm =      58689.9 NCalls =      1
VariableMetric: Iteration #   1 - FCN =    3305420.06058 Edm =      58689.9 NCalls =     12
Info: VariableMetricBuilder: no improvement in line search
Info: VariableMetricBuilder: iterations finish without convergence.
Info in VariableMetricBuilder : edm = 58689.9
Info in             requested : edmval = 0.001
Info in VariableMetricBuilder: INVALID function minimum - edm is above tolerance, : edm = 58689.9
Info in VariableMetricBuilder: Required tolerance  is 10 x edmval  : edmval = 0.001```

**Comment by HDembinski** [comment]
Thanks for the praise. 👍 I didn't change anything that should affect how Migrad converges. If there is a difference, it should come from a possible update of the Minuit C++ code. You could git diff the old and the tag to see whether there was a change in the C++ code. We don't update the C++ code very often, but there was a switch at some point from an internal copy to the standalone repository here https://github.com/GooFit/Minuit2.git.

My hunch is that you are seeing the change in the ncalls keyword in Minuit.migrad. The call limited used to be set to 10000 calls flat. I changed this to use the original MINUIT heuristic, which takes into account that fits with many parameters need more calls. Your fit may run longer before it exhausts the call limit.

The fit really looks very bad, there seem to be serious numerical issues. The EDM value is huge.

**Comment by HDembinski** [comment]
@mayou36 Can I close this?

**Comment by mayou36** [comment]
(sorry, forgot to reply, thanks for the ping!) So I tried to investigate this: interestingly, it is not the `ncalls`, I set this explicitly to 10k (as before).
The fit instability seems to originate from a wrong gradient, it is calculated externally. With the internal gradient, it works smooth.

So I've tried, and it seems that the old version stopped before reaching ncalls maximum while the new one goes on. 

Minimal example (test with 1.3.10 and 1.4.x):
```
from iminuit import Minuit
import numpy as np


def f(x, y, z):
    return (x - 2) ** 2 + (y - 3) ** 2 + (z - 4) ** 2


m = Minuit(f, x=3, y=2.3, z=4.1,
           grad=lambda x, y, z: np.random.uniform([-1] * 3, [1] * 3),
           print_level=3)

m.migrad(ncall=10000)  # run optimiser
print(m.values)  # {'x': 2,'y': 3,'z': 4}

```

What I've found:
- the ncalls argument seemed to be used as a cap, not as a "do this many calls" in 1.3.10, the default heuristic in iminuit seems to be used and it terminates about 55 calls in the above example, no matter what ncalls we give
- in 1.4+, the ncalls will be used as "how much calls to do before failing", if no ncalls is given, the internal heuristic, as above, will be used.

Was ncalls simply ignored in 1.3.10 and always the internal heuristics used?

Proposal: instead of 0, could we also pass None as the default argument not given? I guess it ain't simple and nice to do that with the bindings? Just a "nice to have feature".


Sidenote on the verbosity: The new informations are great! However, I was wondering, can't we have this in a different level? e.g. level 3 is as before and level 4 is the current level 3? (But I can also open a new issue).
